benchmark/src/benchmark/measure.clj

```diff
@@ -11,84 +12,106 @@
+                          (partition 100000))]
+      (d/transact conn (vec entities)))
+    []))
+
+(defn time-context-map
+  ([t config function entity-count datom-count]
+   (time-context-map t config function entity-count datom-count nil))
+  ([t config function entity-count datom-count exec]
+   (cond-> {:time t
+            :context {:dh-config config
+                      :function function
+                      :db-entities entity-count
+                      :db-datoms datom-count}}
+     (some? exec) (assoc-in [:context :execution] exec))))
+
+(defn measure-connection-time [iterations unique-cfg simple-cfg entity-count datom-count]
+  (log/info (str "Measuring connection time on database with config named '" (:name simple-cfg) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (for [_ (range iterations)]
+    (let [timed-conn (timed (d/connect unique-cfg))]
+      (d/release (:res timed-conn))
+      (time-context-map (:t timed-conn) simple-cfg :connection entity-count datom-count))))
+
+(defn measure-query-times
+  [{:keys [iterations data-found-opts query data-types]} initial-tx conn config entity-count datom-count]
+  (log/info (str "Measuring query times on database with config named '" (:name config) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (let [data-found (case data-found-opts
+                     true [true]
+                     false [false]
+                     :all [true false])
+        queries (if (pos? (count initial-tx))
+                  (c/all-queries @conn initial-tx data-types data-found)
+                  (c/non-var-queries @conn data-types (count initial-tx)))
+        filtered-queries (if (= query :all)
+                           queries
+                           (filter #(= query (:function %)) queries))]
+    (flatten
+     (for [{:keys [function query details]} filtered-queries]
+       (do
+         (log/debug (str " Querying with " function " using " details "..."))
+         (for [_ (range iterations)]
+           (time-context-map (:t (timed (d/q query @conn)))
+                             config function entity-count datom-count details)))))))
+
+(defn measure-transaction-times
+  [{:keys [iterations tx-entity-counts]} conn config entity-count datom-count]
+  (log/info (str "Measuring transaction times on database with config named '" (:name config) ", "
+                 (count c/schema) " attributes in entity, and " entity-count " entities in database"))
+  (for [_ (range iterations)    ;; sampling for given entity-count
+        tx-entities tx-entity-counts
+        _ (range iterations)]   ;; sampling for each tx-entities
+    (let [tx (vec (repeatedly tx-entities (partial c/rand-entity Integer/MAX_VALUE)))
+          timed-transact (timed (d/transact conn tx))]
+      (d/transact conn (mapv (fn [dat] [:db.fn/retractEntity (.-e dat)])
```

Comment:
`retract` only cleans up the database if the history is disabled. With history, it should be `purge`. So far, we are not testing configurations with history yet, but we should do that at some point. To avoid confusion and incorrect values in the future, we should already consider both options here.

Reply:
Ah, I see. I used `retract` to avoid deleting and re-creating the database when avoidable—does it matter whether `retract` or `purge` is used in this case?

Reply:
At the moment, purge [only works on databases with history](https://github.com/replikativ/datahike/blob/97eb463b8c71c60f9ff261db795ee679e4e78767/src/datahike/db.cljc#L2078). IMO it would be more intuitive if it would work with both configurations, but I guess that is a discussion on its own.

Reply:
This should work